relay-server/src/actors/upstream.rs

```diff
@@ -276,161 +292,226 @@ impl UpstreamRelay {
-            // futures hanging around, waiting for an open connection). We need to adjust this
-            // timeout to prevent the queued events from timing out while waiting for a free
-            // connection in the pool.
-            //
-            // This may not be good enough in the long run. Right now, filling up the "request
-            // queue" means that requests unrelated to `store` (queries, proxied/forwarded requests)
-            // are blocked by store requests. Ideally, those requests would bypass this queue.
-            //
-            // Two options come to mind:
-            //   1. Have own connection pool for `store` requests
-            //   2. Buffer up/queue/synchronize events before creating the request
             .wait_timeout(self.config.event_buffer_expiry())
             .conn_timeout(self.config.http_connection_timeout())
             // This is the timeout after wait + connect.
             .timeout(self.config.http_timeout())
             .track(ctx.address().recipient())
             .map_err(UpstreamRequestError::SendFailed)
             .and_then(handle_response)
-            .then(|x| {
-                response_sender.send(x).ok();
-                Ok(())
-            });
-        Box::new(future)
+            .into_actor(self)
+            .then(|send_result, slf, ctx| {
+                slf.handle_http_response_status(request, send_result, ctx)
+                    .into_actor(slf)
+            })
+            .spawn(ctx);
+    }
+
+    /// Checks what happened to the request and takes appropriate action
+    ///
+    /// - if the request was sent, notify the response sender
+    /// - if the request was not send:
+    ///     - if it was a network error
+    ///         - if request should be retried
+    ///             - schedule a retry
+    ///         - if there is no scheduled Authentication request
+    ///             - schedule authentication (with backoff)
+    ///     - if it was a non recoverable error, notify the response sender with the error
+    fn handle_http_response_status(
+        &mut self,
+        request: UpstreamRequest,
+        send_result: Result<ClientResponse, UpstreamRequestError>,
+        ctx: &mut Context<Self>,
+    ) -> impl Future<Item = (), Error = ()> {
+        if let Err(err) = send_result {
+            if err.is_network_error() {
+                self.handle_network_error(ctx);
+                if request.retry {
+                    self.enqueue(request, ctx, EnqueuePosition::Back);
+                }
+                return futures::future::failed(());
+            }
+            // we only retry network errors, forward this error and finish
+            request.response_sender.send(Err(err)).ok();
+        } else {
+            // reset any previously failed status
+            // TODO should we reset on success or on anything that is not a network error ?
```

Comment:
This should reset on anything that is not a network error.

Reply:
Ok